### PR TITLE
#275: reconstruct: add new reconstruction for tuple elements with specialization

### LINKED ISSUE #275: Define reconstruction for Pair and Tuple in terms of reconstruction of its elements
A pair/tuple should be as reconstructible as its members, either by default construction or tagged

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #275

### CODE REVIEW

**Comment by lifflander** [comment]
Hmm, we need C++17 enabled here

**Comment by PhilMiller** [comment]
I think this is all good, but I want to look at it more closely before approving

File: src/checkpoint/dispatch/reconstructor.h [hunk 1]
```diff
@@ -51,7 +51,12 @@
 #include <tuple>
 #include <cstdlib>
 
-namespace checkpoint { namespace dispatch {
+namespace checkpoint {
+
+template <typename T>
+struct CheckpointReconstructor;
```
Comment (PhilMiller) [inline src/checkpoint/dispatch/reconstructor.h]: I don't think you need to repeat the declaration here - it's already included.

Comment (PhilMiller) [reply src/checkpoint/dispatch/reconstructor.h]: Other than this, looks good to me.

**Comment by PhilMiller** [comment]
nvcc 10 is definitely not going to work anymore, since it has no support for C++17. For nvcc 11, I think we may need a newer CMake version so that it gets passed `-std=c++17` and not `-std=gnu++1z`

File: src/checkpoint/dispatch/reconstructor.h (hunk 1 above)
Comment (lifflander) [reply src/checkpoint/dispatch/reconstructor.h]: Fixed

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**